pkg/cloudprovider/cloudapi/azure/azure_security.go

```diff
@@ -344,100 +336,172 @@ func (computeCfg *computeServiceConfig) buildEffectiveNSGSecurityRulesToApply(ap
+	if err != nil {
+		return []*armnetwork.SecurityRule{}, err
 	}
-
-	return rules, nil
-}
-
-// buildEffectivePeerNSGSecurityRulesToApply prepares the update rule cloud api payload from internal rules that require peering.
-func (computeCfg *computeServiceConfig) buildEffectivePeerNSGSecurityRulesToApply(appliedToGroupID *securitygroup.CloudResourceID,
-	ingressRules []*securitygroup.CloudRule, egressRules []*securitygroup.CloudRule, perVnetAppliedToNsgName string,
-	rgName string, ruleIP *string) ([]*armnetwork.SecurityRule, error) {
-	// get current rules for applied to SG azure NSG
-	nsgObj, err := computeCfg.nsgAPIClient.get(context.Background(), rgName, perVnetAppliedToNsgName, "")
+	rmEgressSecurityRules, err := convertEgressToNsgSecurityRules(appliedToGroupID, rmERule,
+		agAsgMapByNepheControllerName, atAsgMapByNepheControllerName)
 	if err != nil {
 		return []*armnetwork.SecurityRule{}, err
 	}
 
-	if nsgObj.Properties == nil {
-		return nil, fmt.Errorf("properties field empty in nsg object %s", *nsgObj.ID)
-	}
-	var currentNsgIngressRules []armnetwork.SecurityRule
-	var currentNsgEgressRules []armnetwork.SecurityRule
+	var currentNsgIngressRules []*armnetwork.SecurityRule
+	var currentNsgEgressRules []*armnetwork.SecurityRule
 	currentNsgSecurityRules := nsgObj.Properties.SecurityRules
 	appliedToGroupNepheControllerName := appliedToGroupID.GetCloudName(false)
-	azurePluginLogger().Info("Building peering security rules", "applied to security group", appliedToGroupNepheControllerName)
+	azurePluginLogger().Info("Building security rules", "applied to security group", appliedToGroupNepheControllerName)
 	for _, rule := range currentNsgSecurityRules {
 		if rule.Properties == nil || *rule.Properties.Priority == vnetToVnetDenyRulePriority {
 			continue
 		}
 
-		// Nephe rules will be created from ruleStartPriority and have description.
+		// check if the rule is Nephe rules based on ruleStartPriority and description.
 		desc, ok := securitygroup.ExtractCloudDescription(rule.Properties.Description)
 		if *rule.Properties.Priority >= ruleStartPriority && ok {
-			// skip any rules created by current processing appliedToGroup (as we have new rules for this group).
-			ruleAddrGroupName := desc.AppliedToGroup
+			// check if the rule is created by current processing appliedToGroup.
 			_, _, isNepheControllerCreatedRule := securitygroup.IsNepheControllerCreatedSG(desc.AppliedToGroup)
-			if isNepheControllerCreatedRule && strings.Compare(ruleAddrGroupName, appliedToGroupNepheControllerName) == 0 {
-				continue
+			if isNepheControllerCreatedRule && strings.Compare(desc.AppliedToGroup, appliedToGroupNepheControllerName) == 0 {
+				// skip the rule if found in remove list.
+				removeRules := rmIngressSecurityRules
+				if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionOutbound {
+					removeRules = rmEgressSecurityRules
+				}
+				idx, found := findSecurityRule(rule, removeRules)
```

Comment:
here you are comparing full rules..In your internal CloudRuleIndexer, cannot you store priority or ID for AWS and just skip them for removal?

Reply:
Changed to normalize rule and use deepequal to simplify this logic. Indexer change is a larger change and should have a separate PR.